Fix Subscriptions flows incompatible with HPOS

cesarcosta99 opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #2531 

## Changes proposed in this Pull Request:

Replace all post meta usage with subscription methods that deal with the meta tags properly regardless HPOS is enabled or not.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

1. Set up a store with WooCommerce Subscriptions 4.8.0, latest WooCommerce and Stripe in this branch.
2. [Enable HPOS](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Enabling-High-Performance-Order-Storage-(HPOS)).
3. Test and confirm the following flows work as expected:
  - [Shopper - Subscriptions - Change payment method to new card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#change-payment-method-to-new-card).
  - [Shopper - Subscriptions - Change payment method to saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#change-payment-method-to-saved-card).
  - [Shopper - Subscriptions - Set default payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#set-default-payment-method).
  - [Shopper - Subscriptions - Change default payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#change-default-payment-method).

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

cesarcosta99 committed:
Replace post_meta methods with instance calls

mattallan commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -557,15 +561,16 @@ public function maybe_render_subscription_payment_method( $payment_method_to_dis
 
 				// Take this opportunity to update the key name.
-				update_post_meta( $subscription->get_parent_id(), '_stripe_source_id', $stripe_source_id );
+				$parent_order->update_meta_data( '_stripe_source_id', $stripe_source_id );
```
Missing a `$parent_order->save()` after updating the meta.

mattallan commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -430,8 +432,8 @@ public function delete_renewal_meta( $renewal_order ) {
-		update_post_meta( $subscription->get_id(), '_stripe_customer_id', $renewal_order->get_meta( '_stripe_customer_id', true ) );
-		update_post_meta( $subscription->get_id(), '_stripe_source_id', $renewal_order->get_meta( '_stripe_source_id', true ) );
+		$subscription->update_meta_data( '_stripe_customer_id', $renewal_order->get_meta( '_stripe_customer_id', true ) );
+		$subscription->update_meta_data( '_stripe_source_id', $renewal_order->get_meta( '_stripe_source_id', true ) );
```
Missing a `$subscription->save();` after updating meta data.

mattallan commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -397,12 +399,12 @@ public function maybe_update_source_on_subscription_order( $order, $source ) {
+		$resubscribe_order->delete_meta_data( '_stripe_card_id' );
 		// Delete payment intent ID.
-		delete_post_meta( $resubscribe_order->get_id(), '_stripe_intent_id' );
+		$resubscribe_order->delete_meta_data( '_stripe_intent_id' );
```
Missing a `$resubscribe_order->save()` at the end of this function

cesarcosta99 committed:
Add missing save calls

cesarcosta99 committed:
Add changelog entry

asumaran reviewed (approved):
LGTM. Tests well!

reykjalin reviewed (approved):
I think the changes here all look and test great, but there's a significant amount of function calls still using post meta directly:

```
$ grep -r '_post_' includes

# Unfiltered list, there is some noise here

includes/admin/class-wc-stripe-privacy.php:							'value' => get_post_meta( $order->get_id(), '_stripe_source_id', true ),
includes/admin/class-wc-stripe-privacy.php:							'value' => get_post_meta( $order->get_id(), '_stripe_customer_id', true ),
includes/admin/class-wc-stripe-privacy.php:							'value' => get_post_meta( $subscription->get_id(), '_stripe_source_id', true ),
includes/admin/class-wc-stripe-privacy.php:							'value' => get_post_meta( $subscription->get_id(), '_stripe_customer_id', true ),
includes/admin/class-wc-stripe-privacy.php:		$stripe_source_id = get_post_meta( $subscription_id, '_stripe_source_id', true );
includes/admin/class-wc-stripe-privacy.php:			delete_post_meta( $renewal_order_id, '_stripe_source_id' );
includes/admin/class-wc-stripe-privacy.php:			delete_post_meta( $renewal_order_id, '_stripe_refund_id' );
includes/admin/class-wc-stripe-privacy.php:			delete_post_meta( $renewal_order_id, '_stripe_customer_id' );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $subscription_id, '_stripe_source_id' );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $subscription_id, '_stripe_refund_id' );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $subscription_id, '_stripe_customer_id' );
includes/admin/class-wc-stripe-privacy.php:		$stripe_source_id   = get_post_meta( $order_id, '_stripe_source_id', true );
includes/admin/class-wc-stripe-privacy.php:		$stripe_refund_id   = get_post_meta( $order_id, '_stripe_refund_id', true );
includes/admin/class-wc-stripe-privacy.php:		$stripe_customer_id = get_post_meta( $order_id, '_stripe_customer_id', true );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $order_id, '_stripe_source_id' );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $order_id, '_stripe_refund_id' );
includes/admin/class-wc-stripe-privacy.php:		delete_post_meta( $order_id, '_stripe_customer_id' );
includes/admin/class-wc-stripe-inbox-notes.php:	const POST_SETUP_SUCCESS_ACTION    = 'wc_stripe_apple_pay_post_setup_success';
includes/class-wc-stripe-helper.php:		delete_post_meta( $order_id, self::META_NAME_FEE );
includes/class-wc-stripe-helper.php:		delete_post_meta( $order_id, self::LEGACY_META_NAME_FEE );
includes/class-wc-stripe-helper.php:		delete_post_meta( $order_id, self::META_NAME_NET );
includes/class-wc-stripe-helper.php:		delete_post_meta( $order_id, self::LEGACY_META_NAME_NET );
includes/abstracts/abstract-wc-stripe-connect-rest-controller.php:		add_filter( 'rest_post_dispatch', [ $this, 'send_nocache_header' ], PHP_INT_MAX, 2 );
includes/payment-methods/class-wc-stripe-payment-request.php:		return is_product() || wc_post_content_has_shortcode( 'product_page' );
includes/payment-methods/class-wc-stripe-payment-request.php:		} elseif ( wc_post_content_has_shortcode( 'product_page' ) ) {
includes/class-wc-gateway-stripe.php:		clean_post_cache( $order->get_id() );
```

I don't think we should fix all of these in this PR, but maybe we should open a new PR per-file?

Also not sure how to test these, or if these are even a problem. @cesarcosta99 @asumaran maybe you've already noticed these and determined the leftovers are not an issue?

Commit:
Fix wc_get_orders call in WC_Stripe_Helper

The previous implementation didn't actually filter the meta data, we
were just lucky that the first order in the list fetched happened to
be the one we needed. This commit fixes that by making sure the
metadata filter is added correctly.

Commit:
Fix the rest of the wc_get_orders calls

cesarcosta99 commented:
>Also not sure how to test these, or if these are even a problem.

Thanks for flagging this, @reykjalin. I can't say for sure because we'd need to trace the calls, but I'd expect that issues would have been caught by our extensive testing while going through the critical flows.

On the other hand, I'd like to be extra safe too and have those updated to be in sync with the newest implementations. I created #2538 to pull this into a new issue/PR.
